File: apps-microservices/crawler-monitor-backend/src/lib/systemStats.js
/**
 * System stats helpers — aggregated counters over time windows.
 *
 * Sources:
 *  - Jobs: Redis MGET on crawl_job:* keys (filter by start_time)
 *  - Capacity: optional, from capacity:history:zset
 */

import { readCapacityHistory } from './capacityHistory.js';

const WINDOW_MAP = {
  '1h': 3600000,
  '24h': 86400000,
  '7d': 604800000,
};

export function parseStatsWindow(input) {
  if (typeof input !== 'string' || !(input in WINDOW_MAP)) {
    throw new Error("Invalid window. Use '1h', '24h' or '7d'.");
  }
  return WINDOW_MAP[input];
}

/**
 * Aggregate metrics from a list of jobs that started within `windowMs`.
 * Pure: no side effects, easy to test.
 */
export function aggregateJobStats(jobs, nowMs, windowMs) {
  const cutoff = nowMs - windowMs;
  const inWindow = jobs.filter(j => {
    const t = Date.parse(j.start_time);
    return Number.isFinite(t) && t >= cutoff;
  });

  const counts = { finished: 0, failed: 0, running: 0, archived: 0, restarting_oom: 0, stopping: 0, other: 0 };
  let oomTotal = 0;
  let updateMode = 0;
  let durationsMs = []; // for finished jobs only

  for (const j of inWindow) {
    const status = (j.status || 'other').toLowerCase();
    if (status in counts) counts[status]++;
    else counts.other++;
    oomTotal += j.oom_restart_count || 0;
    if (j.crawl_mode === 'update') updateMode++;
    if (status === 'finished' && j.start_time && j.end_time) {
      const d = Date.parse(j.end_time) - Date.parse(j.start_time);
      if (Number.isFinite(d) && d >= 0) durationsMs.push(d);
    }
  }

  const total = inWindow.length;
  const completed = counts.finished + counts.failed; // jobs with terminal status (excludes archived/running)
  const successRate = completed > 0 ? counts.finished / completed : null;

  const avgDurationMs = durationsMs.length > 0
    ? Math.round(durationsMs.reduce((a, b) => a + b, 0) / durationsMs.length)
    : null;

  return {
    total,
    counts,
    success_rate: successRate,            // 0..1 or null if no terminal jobs
    avg_duration_ms: avgDurationMs,        // null if no finished jobs
    oom_restarts_total: oomTotal,
    update_mode_count: updateMode,
  };
}

/**
 * Aggregate saturation stats from capacity history points.
 * Returns saturated_seconds (sum of intervals where full was true) and saturated_pct.
 */
export function aggregateSaturation(points, windowMs) {
  if (!points || points.length < 2) {
    return { saturated_seconds: 0, saturated_pct: null };
  }
  let saturatedMs = 0;
  for (let i = 1; i < points.length; i++) {
    const prev = points[i - 1];
    const curr = points[i];
    if (prev.full) {
      const dt = curr.ts - prev.ts;
      if (dt > 0) saturatedMs += dt;
    }
  }
  return {
    saturated_seconds: Math.round(saturatedMs / 1000),
    saturated_pct: windowMs > 0 ? saturatedMs / windowMs : null,
  };
}

/**
 * High-level: load jobs + capacity history and compute everything for a single window.
 * `loadJobs(client) -> Promise<Job[]>` is injected for testability.
 */
export async function computeSystemStats(client, windowMs, { loadJobs } = {}) {
  const now = Date.now();
  const jobs = await loadJobs(client);
  const jobStats = aggregateJobStats(jobs, now, windowMs);

  let saturationStats = { saturated_seconds: 0, saturated_pct: null };
  try {
    const points = await readCapacityHistory(client, windowMs);
    saturationStats = aggregateSaturation(points, windowMs);
  } catch (err) {
    // capacity history is best-effort
  }

  return {
    jobs: jobStats,
    capacity: saturationStats,
  };
}

File: apps-microservices/crawler-monitor-backend/internal/domain/systemstats/systemstats.go
// Package systemstats provides aggregated system metric helpers for time windows.
//
// Sources:
//   - Jobs: Redis MGET on crawl_job:* keys (filter by start_time)
//   - Capacity: optional, from capacity:history:zset
package systemstats

import (
	"errors"
	"math"

	"github.com/Hellopro-fr/crawler-monitor-backend/internal/datetime"
)

// WINDOW_MAP mirrors the JS WINDOW_MAP for system stats.
var windowMap = map[string]int64{
	"1h":  3600000,
	"24h": 86400000,
	"7d":  604800000,
}

// ParseStatsWindow parses "1h", "24h" or "7d" and returns the duration in ms.
func ParseStatsWindow(input string) (int64, error) {
	ms, ok := windowMap[input]
	if !ok {
		return 0, errors.New("Invalid window. Use '1h', '24h' or '7d'.")
	}
	return ms, nil
}

// StatusCounts mirrors the counts object in aggregateJobStats.
type StatusCounts struct {
	Finished      int `json:"finished"`
	Failed        int `json:"failed"`
	Running       int `json:"running"`
	Archived      int `json:"archived"`
	RestartingOOM int `json:"restarting_oom"`
	Stopping      int `json:"stopping"`
	Other         int `json:"other"`
}

// JobStats is the output of AggregateJobStats.
type JobStats struct {
	Total           int          `json:"total"`
	Counts          StatusCounts `json:"counts"`
	SuccessRate     *float64     `json:"success_rate"`      // 0..1 or null
	AvgDurationMs   *int64       `json:"avg_duration_ms"`   // null if no finished jobs
	OOMRestartsTotal int         `json:"oom_restarts_total"`
	UpdateModeCount int          `json:"update_mode_count"`
}

// RawJob is a minimal view of a crawl job used for aggregation.
// Fields match Redis job JSON keys (snake_case).
type RawJob struct {
	StartTime       string  `json:"start_time"`
	EndTime         string  `json:"end_time"`
	Status          string  `json:"status"`
	CrawlMode       string  `json:"crawl_mode"`
	OOMRestartCount int     `json:"oom_restart_count"`
}

// AggregateJobStats aggregates metrics from a list of jobs that started within windowMs.
// Pure function — no side effects.
func AggregateJobStats(jobs []RawJob, nowMs, windowMs int64) JobStats {
	cutoff := nowMs - windowMs

	var inWindow []RawJob
	for _, j := range jobs {
		tMs := datetime.ParseStringMs(j.StartTime)
		if tMs < 0 {
			continue
		}
		if tMs >= cutoff {
			inWindow = append(inWindow, j)
		}
	}

	counts := StatusCounts{}
	oomTotal := 0
	updateMode := 0
	var durationsMs []int64

	for _, j := range inWindow {
		status := j.Status
		if status == "" {
			status = "other"
		}
		switch status {
		case "finished":
			counts.Finished++
		case "failed":
			counts.Failed++
		case "running":
			counts.Running++
		case "archived":
			counts.Archived++
		case "restarting_oom":
			counts.RestartingOOM++
		case "stopping":
			counts.Stopping++
		default:
			counts.Other++
		}
		oomTotal += j.OOMRestartCount
		if j.CrawlMode == "update" {
			updateMode++
		}
		if status == "finished" && j.StartTime != "" && j.EndTime != "" {
			stMs := datetime.ParseStringMs(j.StartTime)
			etMs := datetime.ParseStringMs(j.EndTime)
			if stMs >= 0 && etMs >= 0 {
				d := etMs - stMs
				if d >= 0 {
					durationsMs = append(durationsMs, d)
				}
			}
		}
	}

	total := len(inWindow)
	completed := counts.Finished + counts.Failed
	var successRate *float64
	if completed > 0 {
		sr := float64(counts.Finished) / float64(completed)
		successRate = &sr
	}
	var avgDurationMs *int64
	if len(durationsMs) > 0 {
		sum := int64(0)
		for _, d := range durationsMs {
			sum += d
		}
		avg := int64(math.Round(float64(sum) / float64(len(durationsMs))))
		avgDurationMs = &avg
	}

	return JobStats{
		Total:            total,
		Counts:           counts,
		SuccessRate:      successRate,
		AvgDurationMs:    avgDurationMs,
		OOMRestartsTotal: oomTotal,
		UpdateModeCount:  updateMode,
	}
}

// CapacityPoint is a single capacity history snapshot.
type CapacityPoint struct {
	Ts      int64 `json:"ts"`
	Running int   `json:"running"`
	Max     int   `json:"max"`
	Full    bool  `json:"full"`
}

// SaturationStats is the output of AggregateSaturation.
type SaturationStats struct {
	SaturatedSeconds int      `json:"saturated_seconds"`
	SaturatedPct     *float64 `json:"saturated_pct"` // null if windowMs == 0
}

// AggregateSaturation sums intervals where full=true.
// Mirrors JS aggregateSaturation exactly.
func AggregateSaturation(points []CapacityPoint, windowMs int64) SaturationStats {
	if len(points) < 2 {
		return SaturationStats{SaturatedSeconds: 0, SaturatedPct: nil}
	}
	var saturatedMs int64
	for i := 1; i < len(points); i++ {
		prev := points[i-1]
		curr := points[i]
		if prev.Full {
			dt := curr.Ts - prev.Ts
			if dt > 0 {
				saturatedMs += dt
			}
		}
	}
	secs := int(math.Round(float64(saturatedMs) / 1000.0))
	var pct *float64
	if windowMs > 0 {
		p := float64(saturatedMs) / float64(windowMs)
		pct = &p
	}
	return SaturationStats{
		SaturatedSeconds: secs,
		SaturatedPct:     pct,
	}
}

// SystemStatsResult is the combined output for /api/system/stats.
type SystemStatsResult struct {
	Jobs     JobStats        `json:"jobs"`
	Capacity SaturationStats `json:"capacity"`
}
